File: go/1110. Delete Nodes And Return Forest.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func delNodes(root *TreeNode, to_delete []int) []*TreeNode {
    delMap := make(map[int]bool)
    for _, v := range to_delete {
        delMap[v] = true
    }
    return helper(root, nil, delMap, true)
}
func helper(root, parent *TreeNode, delMap map[int]bool, leftChild bool) []*TreeNode {
    result := []*TreeNode{}
    if root == nil {
        return result
    }
    if delMap[root.Val] {
        result = append(result, helper(root.Left, nil, delMap, true)...)
        result = append(result, helper(root.Right, nil, delMap, false)...)
        if parent != nil {
            if leftChild {
                parent.Left = nil
            } else {
                parent.Right = nil
            }
        }
    } else {
        if parent == nil {
            result = append(result, root)
        }
        result = append(result, helper(root.Left, root, delMap, true)...)
        result = append(result, helper(root.Right, root, delMap, false)...)
    }
    return result
}

File: java/1110. Delete Nodes And Return Forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] toDelete) {
        List<TreeNode> result = new ArrayList<>();
        Set<Integer> set = new HashSet<>();
        for (int i : toDelete) set.add(i);
        helper(root, null, set, result);
        return result;
    }
    private void helper(TreeNode root, TreeNode previous, Set<Integer> toDelete, List<TreeNode> result) {
        if (root == null) return;
        if (previous == null) {
            if (toDelete.contains(root.val)) {
                helper(root.left, null, toDelete, result);
                helper(root.right, null, toDelete, result);
                root.left = null;
                root.right = null;
            } else {
                result.add(root);
            }
        }
        if (root.left != null && toDelete.contains(root.left.val)) {
            helper(root.left, null, toDelete, result);
            root.left = null;
        } else {
            helper(root.left, root, toDelete, result);
        }
        if (root.right != null && toDelete.contains(root.right.val)) {
            helper(root.right, null, toDelete, result);
            root.right = null;
        } else {
            helper(root.right, root, toDelete, result);
        }
    }
}